### ISSUE TITLE
<clearDefaultOptions> also clears 'internal' implied options which are added automatically and essential for the build

### ISSUE THREAD

**Issue by fukanchik** [maven-nar/nar-maven-plugin#148]
amd64-Linux-gpp
when &lt;clearDefaultOptions>true&lt;/clearDefaultOptions> is used, the following implied options are cleared as well - making the build impossible, since for example without '-c' gpp would try to generate application and would fail since shared/static library would not have 'main()':
- '-fPIC'
- '-c'

please note that the options above were not part of aol.properties but were added by nar automagically in addImpliedArgs(...)

**Comment by dproteus13** [comment]
How would everyone feel about a solution that kind of fit in the middle?  I'm thinking of a way to remove a specific option from the list:
```xml
<clearOptions>
<clearOption>-wall</clearOption>
</clearOptions>

something like this make sense?  That way you can remove those 'default' options that you don't want, without breaking your ability to build?

**Comment by dproteus13** [comment]
From looking at this very briefly today, I _think_ this would be done in `cpptasks/compiler/CommandLineCompiler.java::createConfiguration()`, right after the block of code that says `add all appropriate defines and undefines` in its comment.  This way we build up all of the options normally, and then remove those options the user has asked us to clear.

Does this approach make sense?  Does anyone have any additional pointers for someone to go about implementing this?  I'm hoping to try my hand...

**Comment by dscho** [comment]
My gut feeling is that changing the `<clearDefaultOptions>` from a boolean to a tristate (`true`, `false`, `all`) is the best line of attack.

**Comment by dproteus13** [comment]
Okay, then true would change meaning to only remove the options from the AOL?  Or would `{true, false, partial}` be cleaner to maintain backwards compatibility?

I hope I'm not being obnoxious, I'm just excited to actually get to (hopefully) contribute finally, and want to get it right :)

**Comment by dscho** [comment]
> ... then true would change meaning...

My idea of extending the current behavior was to keep the existing functionality as it is, to allow existing users [not having to change anything in their setup](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

**Comment by dproteus13** [comment]
I absolutely agree with POLA, but I thought @fukanchik 's original issue was that setting `<clearDefaultOptions>true</clearDefaultOptions>` cleared 'all' options.  So making it a tristate of `{true, false, all}` seems like either `all` will be redundant or `true` will change meaning.  Or am I misinterpreting the original issue report?

**Comment by dproteus13** [comment]
So I thought about your response a little more and think I realize now what you were getting at...  Since the issue was that the 'true' state didn't do the thing that the users expect, fix that.  If I'm now understanding right, then your `{true, false, all}` makes the most sense.  Since it was what you suggested originally, I'll try to start on it unless you disagree :)

**Comment by dscho** [comment]
> Since the issue was that the 'true' state didn't do the thing that the users expect, fix that.

No, what I was saying is: we would not know whether people rely on the "wrong" behavior, therefore it would be a backwards-incompatible modification to change that behavior now.

So what I was suggesting instead is to keep the behavior with the originally possible values `true` and `false` exactly as-is, and just add a new value `all` that clears all default options, including the options that the individual compiler back ends want to hard-code. That way, you won't break existing users' projects but you still get the behavior this ticket was asking for.

**Comment by dscho** [comment]
Actually, I seem to have misread the original report. From my vague memory of reading that part of the code, I was under the impression that `<clearDefaultOptions>true</clearDefaultOptions>` cleared out only *part* of the options.

So maybe the best idea would be to start by writing a unit test that illustrates the desired behavior so that we can be on the same page as to what we all want and can agree on.
